### PR TITLE
Unification support for inverting hide/reveal

### PR DESCRIPTION
**Pull request by nikswamy**
There are two new rules in the unifier providing special treatment for FStar.Ghost.hide/reveal

```
                // 1.
                // reveal ?u == y, where head y <> hide/reveal
                //   by generating hide (reveal ?u) == hide y
                //   and simplifying it to       ?u == hide y
                //
                // 2.
                //  hide ?u == y, where head y <> hide/reveal
                //  by generating reveal (hide ?u) == reveal y
                //  and simplifying it to       ?u == reveal y
```            
   
This allows the following code to work
  
```
assume
val t (x:int) : Type0
assume
val f (#x:erased int) (y:t (reveal x)) : unit
let test1 (y:t 17) : unit = f y //F* can now infer the implicit arg here, as below
let test1' (y:t 17) : unit = f #(hide 17) y

assume
val s (x:erased int) : Type0
assume
val g (#x:int) (y: t (hide x)) : unit
let test2 (i:erased int) (y:t i) : unit = g y //F* can now infer the implicit arg here, as below
let test2' (i:erased int) (y:t i) : unit = g #(reveal i) y
```  

Motivated by use cases in Steel where this limitation of type inference kept coming up.

### CODE REVIEW

File: src/typechecker/FStar.TypeChecker.Env.fs [hunk 1]
```diff
@@ -2017,3 +2021,12 @@ let get_letrec_arity (env:env) (lbname:lbname) : option<int> =
                     env.letrecs with
   | Some (_, arity, _, _) -> Some arity
   | None -> None
+
+let fvar_of_nonqual_lid env lid =
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Env.fs:2025]: This is just a utility to build an fvar with the delta_depth set properly

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 2]
```diff
@@ -2809,6 +2882,12 @@ and solve_t' (env:Env.env) (problem:tprob) (wl:worklist) : solution =
               solve env wl
 
             | Inr None ->
+
+              match try_reveal_hide env t1 t2 with
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs:2886]: In the rigid/rigid case, if there's a head mismatch, try the reveal/hide rules

**Comment by nikswamy** [comment]
Discussed separately with Aseem and Aymeric, thanks for the feedback.

**Comment by R1kM** [comment]
As discussed on Slack, this looks good to me.

**Pull request merged**